File: chapters/ch05_microgpt/scripts/gmr_clean.py
#!/usr/bin/env python3
"""GMR cleaning for microGPT-style corpus prep (pure Python).

This script implements Geometric Manifold Rectification (GMR) as a
preprocessing stage before language-model training.

Input format:
- docs file: one document per line
- labels file: one label per line (line-aligned with docs)
  - supports 0/1 or aliases like majority/minority, neg/pos

Pipeline:
1) bytes_tokens(doc) -> list[int]
2) hashed_bow_embedding(token_ids) -> dense vector
3) gmr_keep_mask(X, y) -> keep/drop per document
4) write kept docs to output file

Usage:
  python chapters/ch05_microgpt/scripts/gmr_clean.py \
      --docs docs.txt --labels labels.txt --output input_clean.txt
"""

from __future__ import annotations

import argparse
import heapq
import math
from pathlib import Path


def bytes_tokens(text: str) -> list[int]:
    return list(text.encode("utf-8", errors="ignore"))


def hashed_bow_embedding(token_ids: list[int], dim: int = 256) -> list[float]:
    vec = [0.0] * dim
    for token in token_ids:
        vec[token % dim] += 1.0

    sq = 0.0
    for value in vec:
        sq += value * value
    norm = math.sqrt(sq) + 1e-12
    for i in range(dim):
        vec[i] /= norm
    return vec


def _dot(a: list[float], b: list[float]) -> float:
    total = 0.0
    for i in range(len(a)):
        total += a[i] * b[i]
    return total


def _norm(a: list[float]) -> float:
    return math.sqrt(_dot(a, a))


def _euclidean(a: list[float], b: list[float]) -> float:
    sq = 0.0
    for i in range(len(a)):
        delta = a[i] - b[i]
        sq += delta * delta
    return math.sqrt(sq)


def _cosine_dist(
    a: list[float], b: list[float], na: float | None = None, nb: float | None = None
) -> float:
    eps = 1e-12
    if na is None:
        na = _norm(a)
    if nb is None:
        nb = _norm(b)
    return 1.0 - (_dot(a, b) / (na * nb + eps))


def gmr_keep_mask(
    x: list[list[float]],
    y: list[int],
    k: int = 15,
    alpha: float = 0.3,
    beta: float = 0.7,
    gamma: float = 0.1,
    metric: str | None = None,
) -> list[bool]:
    """Pure-Python GMR implementation.

    Args:
      x: Dense vectors
      y: Binary labels (0=majority, 1=minority)
      k, alpha, beta, gamma: GMR hyperparameters
      metric: "euclidean", "cosine", or None (auto by dimension)
    """

    n = len(x)
    if n == 0:
        return []

    d = len(x[0])
    n_min = 0
    for label in y:
        if label == 1:
            n_min += 1

    if n_min < 10:
        return [True] * n

    if metric is None:
        metric = "cosine" if d > 100 else "euclidean"

    norms = None
    if metric == "cosine":
        norms = [_norm(vec) for vec in x]

    def dist(i: int, j: int) -> float:
        if metric == "cosine":
            assert norms is not None
            return _cosine_dist(x[i], x[j], norms[i], norms[j])
        return _euclidean(x[i], x[j])

    kk = min(k, n - 1)
    if kk <= 0:
        return [True] * n

    yhat = [0] * n
    v0 = [0.0] * n
    v1 = [0.0] * n
    eps = 1e-12

    for i in range(n):
        neighbors = heapq.nsmallest(
            kk,
            ((dist(i, j), j) for j in range(n) if j != i),
            key=lambda pair: pair[0],
        )

        weights: list[tuple[float, int]] = []
        wsum = 0.0
        for dij, j in neighbors:
            w = 1.0 / (dij + eps)
            weights.append((w, j))
            wsum += w

        s0 = 0.0
        s1 = 0.0
        inv = 1.0 / (wsum + eps)
        for w, j in weights:
            wn = w * inv
            if y[j] == 1:
                s1 += wn
            else:
                s0 += wn

        v0[i] = s0
        v1[i] = s1
        yhat[i] = 1 if s1 > s0 else 0

    remove = [False] * n

    for i in range(n):
        if y[i] == 0:
            conf = v0[i]
            if yhat[i] == 1 or conf < alpha:
                remove[i] = True

    candidates: list[int] = []
    for i in range(n):
        if y[i] == 1 and yhat[i] == 0 and v0[i] > beta:
            candidates.append(i)

    budget = int(math.floor(gamma * n_min))
    if budget > 0 and candidates:
        candidates.sort(key=lambda i: v0[i], reverse=True)
        for i in candidates[:budget]:
            remove[i] = True

    return [not flag for flag in remove]


def parse_label(raw: str) -> int:
    value = raw.strip().lower()
    if value in {"1", "minority", "min", "pos", "positive", "clean", "good"}:
        return 1
    if value in {"0", "majority", "maj", "neg", "negative", "noisy", "bad"}:
        return 0
    raise ValueError(
        f"Unsupported label '{raw}'. Use 0/1 or majority/minority aliases."
    )


def read_lines(path: Path) -> list[str]:
    return path.read_text(encoding="utf-8").splitlines()


def main() -> None:
    parser = argparse.ArgumentParser(description="GMR cleaner for corpus documents")
    parser.add_argument(
        "--docs", required=True, help="Path to docs file (one line per doc)"
    )
    parser.add_argument(
        "--labels",
        required=True,
        help="Path to labels file (one line per label, aligned)",
    )
    parser.add_argument(
        "--output", default="input_clean.txt", help="Output path for cleaned docs"
    )
    parser.add_argument("--dim", type=int, default=256, help="Embedding dimension")
    parser.add_argument("--k", type=int, default=15, help="GMR k")
    parser.add_argument("--alpha", type=float, default=0.3, help="GMR alpha")
    parser.add_argument("--beta", type=float, default=0.7, help="GMR beta")
    parser.add_argument("--gamma", type=float, default=0.1, help="GMR gamma")
    parser.add_argument(
        "--metric",
        choices=["euclidean", "cosine", "auto"],
        default="auto",
        help="Distance metric",
    )
    args = parser.parse_args()

    docs_path = Path(args.docs)
    labels_path = Path(args.labels)
    out_path = Path(args.output)

    docs = read_lines(docs_path)
    labels_raw = read_lines(labels_path)

    if len(docs) != len(labels_raw):
        raise ValueError(
            f"Line mismatch: docs={len(docs)} labels={len(labels_raw)}. They must align."
        )

    labels = [parse_label(v) for v in labels_raw]

    x: list[list[float]] = []
    for doc in docs:
        token_ids = bytes_tokens(doc)
        x.append(hashed_bow_embedding(token_ids, dim=args.dim))

    metric = None if args.metric == "auto" else args.metric
    keep = gmr_keep_mask(
        x,
        labels,
        k=args.k,
        alpha=args.alpha,
        beta=args.beta,
        gamma=args.gamma,
        metric=metric,
    )

    kept_docs = [doc for doc, flag in zip(docs, keep) if flag]
    kept_labels = [label for label, flag in zip(labels, keep) if flag]

    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(
        "\n".join(kept_docs) + ("\n" if kept_docs else ""), encoding="utf-8"
    )

    kept_count = len(kept_docs)
    dropped_count = len(docs) - kept_count
    kept_min = sum(1 for v in kept_labels if v == 1)
    kept_maj = sum(1 for v in kept_labels if v == 0)

    print("GMR cleaning complete")
    print(f"- input docs: {len(docs)}")
    print(f"- kept docs: {kept_count}")
    print(f"- dropped docs: {dropped_count}")
    print(f"- kept minority/majority: {kept_min}/{kept_maj}")
    print(f"- output: {out_path}")


if __name__ == "__main__":
    main()

File: chapters/ch05_microgpt/src/bin/gmr_clean.rs
//! GMR cleaning for microGPT-style corpus prep (pure Rust).
//!
//! Implements Geometric Manifold Rectification (GMR) as a preprocessing
//! stage before language-model training.
//!
//! Usage:
//!   cargo run -p ch05_microgpt --bin gmr_clean -- \
//!       --docs docs.txt --labels labels.txt --output input_clean.txt

use std::cmp::Ordering;
use std::fs::{self, File};
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;

/// Convert text to byte token IDs.
fn bytes_tokens(text: &str) -> Vec<u32> {
    text.as_bytes().iter().map(|&b| b as u32).collect()
}

/// Create a hashed bag-of-words embedding with L2 normalization.
fn hashed_bow_embedding(token_ids: &[u32], dim: usize) -> Vec<f64> {
    let mut vec = vec![0.0; dim];
    for &token in token_ids {
        vec[(token as usize) % dim] += 1.0;
    }

    // L2 normalize
    let sq: f64 = vec.iter().map(|v| v * v).sum();
    let norm = sq.sqrt().max(1e-12);
    for v in &mut vec {
        *v /= norm;
    }
    vec
}

fn dot(a: &[f64], b: &[f64]) -> f64 {
    a.iter().zip(b.iter()).map(|(x, y)| x * y).sum()
}

fn norm(a: &[f64]) -> f64 {
    dot(a, a).sqrt()
}

fn euclidean(a: &[f64], b: &[f64]) -> f64 {
    a.iter()
        .zip(b.iter())
        .map(|(x, y)| (x - y).powi(2))
        .sum::<f64>()
        .sqrt()
}

fn cosine_dist(a: &[f64], b: &[f64], na: Option<f64>, nb: Option<f64>) -> f64 {
    let na = na.unwrap_or_else(|| norm(a));
    let nb = nb.unwrap_or_else(|| norm(b));
    1.0 - dot(a, b) / (na * nb + 1e-12)
}

/// GMR keep/drop mask computation.
///
/// Args:
///   x: Dense vectors
///   y: Binary labels (0=majority, 1=minority)
///   k, alpha, beta, gamma: GMR hyperparameters
///   metric: "euclidean", "cosine", or None (auto by dimension)
fn gmr_keep_mask(
    x: &[Vec<f64>],
    y: &[i32],
    k: usize,
    alpha: f64,
    beta: f64,
    gamma: f64,
    metric: Option<&str>,
) -> Vec<bool> {
    let n = x.len();
    if n == 0 {
        return vec![];
    }

    let d = x[0].len();
    let n_min: usize = y.iter().filter(|&&label| label == 1).count();

    if n_min < 10 {
        return vec![true; n];
    }

    // Auto-select metric based on dimension
    let metric = match metric {
        Some(m) => m.to_string(),
        None => {
            if d > 100 {
                "cosine".to_string()
            } else {
                "euclidean".to_string()
            }
        }
    };

    // Precompute norms for cosine distance
    let norms: Option<Vec<f64>> = if metric == "cosine" {
        Some(x.iter().map(|v| norm(v)).collect())
    } else {
        None
    };

    let kk = k.min(n - 1);
    if kk == 0 {
        return vec![true; n];
    }

    let mut yhat = vec![0i32; n];
    let mut v0 = vec![0.0f64; n];
    let mut v1 = vec![0.0f64; n];
    let eps = 1e-12;

    for i in 0..n {
        // Find k nearest neighbors
        let mut distances: Vec<(f64, usize)> = (0..n)
            .filter(|&j| j != i)
            .map(|j| {
                let dist = if metric == "cosine" {
                    let norms = norms.as_ref().unwrap();
                    cosine_dist(&x[i], &x[j], Some(norms[i]), Some(norms[j]))
                } else {
                    euclidean(&x[i], &x[j])
                };
                (dist, j)
            })
            .collect();

        // Sort and take k nearest neighbors
        distances.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(Ordering::Equal));
        let neighbors = &distances[..kk];

        // Compute inverse distance weights
        let weights: Vec<(f64, usize)> = neighbors
            .iter()
            .map(|&(dij, j)| (1.0 / (dij + eps), j))
            .collect();

        let wsum: f64 = weights.iter().map(|(w, _)| w).sum();
        let inv = 1.0 / (wsum + eps);

        let mut s0 = 0.0;
        let mut s1 = 0.0;
        for (w, j) in &weights {
            let wn = w * inv;
            if y[*j] == 1 {
                s1 += wn;
            } else {
                s0 += wn;
            }
        }

        v0[i] = s0;
        v1[i] = s1;
        yhat[i] = if s1 > s0 { 1 } else { 0 };
    }

    let mut remove = vec![false; n];

    // Remove low-confidence or misclassified majority samples
    for i in 0..n {
        if y[i] == 0 {
            let conf = v0[i];
            if yhat[i] == 1 || conf < alpha {
                remove[i] = true;
            }
        }
    }

    // Find minority candidates with high majority neighbor weight
    let mut candidates: Vec<usize> = (0..n)
        .filter(|&i| y[i] == 1 && yhat[i] == 0 && v0[i] > beta)
        .collect();

    // Remove up to gamma fraction of minority
    let budget = (gamma * n_min as f64).floor() as usize;
    if budget > 0 && !candidates.is_empty() {
        candidates.sort_by(|&a, &b| v0[b].partial_cmp(&v0[a]).unwrap_or(Ordering::Equal));
        for &i in candidates.iter().take(budget) {
            remove[i] = true;
        }
    }

    remove.into_iter().map(|r| !r).collect()
}

/// Parse label string to binary integer.
fn parse_label(raw: &str) -> Result<i32, String> {
    let value = raw.trim().to_lowercase();
    if ["1", "minority", "min", "pos", "positive", "clean", "good"].contains(&value.as_str()) {
        return Ok(1);
    }
    if ["0", "majority", "maj", "neg", "negative", "noisy", "bad"].contains(&value.as_str()) {
        return Ok(0);
    }
    Err(format!(
        "Unsupported label '{}'. Use 0/1 or majority/minority aliases.",
        raw.trim()
    ))
}

/// Read lines from a file.
fn read_lines(path: &PathBuf) -> Result<Vec<String>, String> {
    let file = File::open(path).map_err(|e| format!("Failed to open {}: {}", path.display(), e))?;
    let reader = BufReader::new(file);
    let mut lines = Vec::new();
    for line in reader.lines() {
        lines.push(line.map_err(|e| format!("Failed to read {}: {}", path.display(), e))?);
    }
    Ok(lines)
}

fn main() -> Result<(), String> {
    let args: Vec<String> = std::env::args().collect();

    let mut docs_path: Option<PathBuf> = None;
    let mut labels_path: Option<PathBuf> = None;
    let mut output_path = PathBuf::from("input_clean.txt");
    let mut dim: usize = 256;
    let mut k: usize = 15;
    let mut alpha: f64 = 0.3;
    let mut beta: f64 = 0.7;
    let mut gamma: f64 = 0.1;
    let mut metric: Option<String> = None;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--docs" => {
                i += 1;
                docs_path = Some(PathBuf::from(args.get(i).ok_or("--docs requires a value")?));
            }
            "--labels" => {
                i += 1;
                labels_path = Some(PathBuf::from(
                    args.get(i).ok_or("--labels requires a value")?,
                ));
            }
            "--output" => {
                i += 1;
                output_path = PathBuf::from(args.get(i).ok_or("--output requires a value")?);
            }
            "--dim" => {
                i += 1;
                dim = args
                    .get(i)
                    .ok_or("--dim requires a value")?
                    .parse()
                    .map_err(|e| format!("Invalid dim: {}", e))?;
            }
            "--k" => {
                i += 1;
                k = args
                    .get(i)
                    .ok_or("--k requires a value")?
                    .parse()
                    .map_err(|e| format!("Invalid k: {}", e))?;
            }
            "--alpha" => {
                i += 1;
                alpha = args
                    .get(i)
                    .ok_or("--alpha requires a value")?
                    .parse()
                    .map_err(|e| format!("Invalid alpha: {}", e))?;
            }
            "--beta" => {
                i += 1;
                beta = args
                    .get(i)
                    .ok_or("--beta requires a value")?
                    .parse()
                    .map_err(|e| format!("Invalid beta: {}", e))?;
            }
            "--gamma" => {
                i += 1;
                gamma = args
                    .get(i)
                    .ok_or("--gamma requires a value")?
                    .parse()
                    .map_err(|e| format!("Invalid gamma: {}", e))?;
            }
            "--metric" => {
                i += 1;
                let m = args.get(i).ok_or("--metric requires a value")?;
                if m != "auto" {
                    metric = Some(m.clone());
                }
            }
            "--help" | "-h" => {
                println!("GMR cleaner for corpus documents");
                println!();
                println!("Usage: cargo run -p ch05_microgpt --bin gmr_clean -- [OPTIONS]");
                println!();
                println!("Options:");
                println!("  --docs PATH      Path to docs file (required)");
                println!("  --labels PATH    Path to labels file (required)");
                println!(
                    "  --output PATH    Output path for cleaned docs [default: input_clean.txt]"
                );
                println!("  --dim N          Embedding dimension [default: 256]");
                println!("  --k N            GMR k neighbors [default: 15]");
                println!("  --alpha F        GMR alpha threshold [default: 0.3]");
                println!("  --beta F         GMR beta threshold [default: 0.7]");
                println!("  --gamma F        GMR gamma fraction [default: 0.1]");
                println!(
                    "  --metric M       Distance metric: euclidean, cosine, auto [default: auto]"
                );
                println!("  --help, -h       Show this help message");
                return Ok(());
            }
            other => {
                return Err(format!("Unknown argument: {}", other));
            }
        }
        i += 1;
    }

    let docs_path = docs_path.ok_or("--docs is required")?;
    let labels_path = labels_path.ok_or("--labels is required")?;

    // Read input files
    let docs = read_lines(&docs_path)?;
    let labels_raw = read_lines(&labels_path)?;

    if docs.len() != labels_raw.len() {
        return Err(format!(
            "Line mismatch: docs={} labels={}. They must align.",
            docs.len(),
            labels_raw.len()
        ));
    }

    // Parse labels
    let labels: Result<Vec<i32>, String> = labels_raw.iter().map(|v| parse_label(v)).collect();
    let labels = labels?;

    // Compute embeddings
    let x: Vec<Vec<f64>> = docs
        .iter()
        .map(|doc| {
            let token_ids = bytes_tokens(doc);
            hashed_bow_embedding(&token_ids, dim)
        })
        .collect();

    // Run GMR
    let metric_arg = metric.as_deref();
    let keep = gmr_keep_mask(&x, &labels, k, alpha, beta, gamma, metric_arg);

    // Filter and write output
    let kept_docs: Vec<&String> = docs
        .iter()
        .zip(keep.iter())
        .filter_map(|(d, &k)| if k { Some(d) } else { None })
        .collect();
    let kept_labels: Vec<i32> = labels
        .iter()
        .zip(keep.iter())
        .filter_map(|(&l, &k)| if k { Some(l) } else { None })
        .collect();

    // Create parent directory if needed
    if let Some(parent) = output_path.parent() {
        fs::create_dir_all(parent)
            .map_err(|e| format!("Failed to create directory {}: {}", parent.display(), e))?;
    }

    let mut file = File::create(&output_path)
        .map_err(|e| format!("Failed to create {}: {}", output_path.display(), e))?;

    for (i, doc) in kept_docs.iter().enumerate() {
        if i > 0 {
            file.write_all(b"\n")
                .map_err(|e| format!("Write error: {}", e))?;
        }
        file.write_all(doc.as_bytes())
            .map_err(|e| format!("Write error: {}", e))?;
    }
    if !kept_docs.is_empty() {
        file.write_all(b"\n")
            .map_err(|e| format!("Write error: {}", e))?;
    }

    let kept_count = kept_docs.len();
    let dropped_count = docs.len() - kept_count;
    let kept_min: usize = kept_labels.iter().filter(|&&l| l == 1).count();
    let kept_maj: usize = kept_labels.iter().filter(|&&l| l == 0).count();

    println!("GMR cleaning complete");
    println!("- input docs: {}", docs.len());
    println!("- kept docs: {}", kept_count);
    println!("- dropped docs: {}", dropped_count);
    println!("- kept minority/majority: {}/{}", kept_min, kept_maj);
    println!("- output: {}", output_path.display());

    Ok(())
}
